sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs

```diff
@@ -220,13 +273,127 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
                     .ok_or(TransportError::InvalidOperation)?,
             );
         }
+
+        if let Some(cmsis_interface) = self.cmsis_interface {
+            // HyperDebug firmware supports binary protocol for retrieving list of events, use
+            // that for greatly improved performance.
+
+            let mut pkt = Vec::<u8>::new();
+            pkt.write_u8(Self::CMSIS_DAP_CUSTOM_COMMAND_GPIO)?;
+            pkt.write_u8(Self::GPIO_MONITORING_READ)?;
+            pkt.write_u8(pin_names.len().try_into()?)?;
+            for pin_name in &pin_names {
+                pkt.write_u8(pin_name.len().try_into()?)?;
+                pkt.extend_from_slice(pin_name.as_bytes());
+            }
+            self.inner
+                .usb_device
+                .borrow()
+                .write_bulk(cmsis_interface.out_endpoint, &pkt)?;
+
+            let mut databytes: Vec<u8> =
+                vec![0u8; 1 + size_of::<RspGpioMonitoringHeader>() + USB_MAX_SIZE];
+            let mut bytecount = 0;
+
+            while bytecount < 1 + size_of::<RspGpioMonitoringHeader>() {
```

Comment:
The header always fits in one packet. Would it make sense to change the code to first do an initial `read_bulk` to get the header + part of the data. With this, we now know the size of the full data so we can allocate a buffer for it and read it with several `read_bulk`?

Reply:
From my interpretation of @jesultra's comment it sounds like they're using USB as a stream and doesn't guarantee the header'll be sent in a single packet?

Reply:
Ah sorry, I missed @jesultra 's comment on the USB protocol.